play-json-tools/src/main/scala/com/evolutiongaming/util/JsonFormats.scala

```diff
@@ -234,10 +234,44 @@ object JsonFormats {
     }
 
     def writes(t: String, json: JsObject = Json.obj()): JsObject = {
+      if (json.keys contains "type") sys error s"Inner JSON for '$t' subtype already contains 'type' field"
       Json.obj("type" -> t) ++ json
     }
   }
 
+  object FoldedTypeFormat {
+
+    def reads[T](readsPf: JsValue => PartialFunction[String, JsResult[T]]): Reads[T] = new Reads[T] {
+      def reads(json: JsValue): JsResult[T] = {
+        def reads(json: JsValue, t: String) = {
+          val pf = readsPf(json)
+          if (pf isDefinedAt t) pf(t)
+          else JsError(s"No Reads defined for $t")
+        }
+
+        for {
+          typ <- (json \ "type").validate[String]
+          result <- typ.split("#").toList match {
+            case head :: Nil  => reads(json, head)
+            case head :: tail =>
+              val j = json.as[JsObject] ++ Json.obj("type" -> tail.mkString("#"))
+              reads(j, head)
+            case Nil          => sys error "It's impossible"
+          }
+        } yield result
+      }
+    }
+
+    def writes[T](writesPf: PartialFunction[T, (String, JsObject)]): OWrites[T] = new OWrites[T] {
```

Comment:
don't you think that passing it will be simpler replacing `writesPf: PartialFunction[T, (String, JsObject)]`
with `writes: T => (String, JsObject)` ?